### PR TITLE
feat: Segmentation model API prototype

### PR DESCRIPTION
**Pull request by intsco**
The PR adds support of a simple segmentation model API where every container has it's own internal job queue, an image storage and a worker thread. All segmentation requests are handled asynchronously after the initial POST request. The results can be fetched with a GET request afterwards using the provided `job_id`.

Details:
* Refactored and split API creation depending on model_type.
* Separate modules for classification and segmentation models.
* `logger` module.
* Segmentation API tests and test data.
* Bumped version.
* `pytest`

### CODE REVIEW

File: telesto/models/classification.py [hunk 1]
```diff
@@ -29,7 +28,7 @@ def _load_model(self, model_path: str):
     def predict(self, input_list: List[np.ndarray]) -> np.ndarray:
         raise NotImplemented
 
-    def _preprocess(self, doc):
+    def _preprocess(self, doc: dict) -> List[np.ndarray]:
```
Comment (koosk) [inline telesto/models/classification.py:31]: I expect these to be the same in `SegmentationModelBase`, but not sure. Maybe this part of the code could be reused.

File: telesto/models/segmentation.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class DummySegmentationModel(SegmentationModelBase):
+    def __init__(self, storage: ImageStorage):
+        super().__init__(classes=["fg", "bg"], model_path='', storage=storage)
+
+    def _load_model(self, model_path: str):
+        pass
+
+    def predict(self, input: np.ndarray) -> np.ndarray:
+        if input.ndim == 3:
+            input = input[:, :, 0]
```
Comment (koosk) [inline telesto/models/segmentation.py:44]: Should it not be a binary (thresholded?) image?

File: telesto/apps/segmentation.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+class SegmentationBase:
+    def on_get(self, req, resp):
+        doc = {"status": "ok", "host": socket.getfqdn(), "worker.pid": os.getpid()}
+        resp.body = json.dumps(doc, ensure_ascii=False)
+
+
+class ImageStorage:
+
+    def __init__(self, base_path: str = "/tmp/telesto/images"):
```
Comment (koosk) [inline telesto/apps/segmentation.py:48]: I'd rather use [tempfile](https://docs.python.org/3/library/tempfile.html) here.

File: telesto/apps/segmentation.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    def save(self, gid: str, image: Image, is_mask: bool):
+        image.save(self._image_path(gid, is_mask), format="PNG")
+
+    def load(self, gid: str, is_mask: bool) -> Image:
+        image_path = self._image_path(gid, is_mask)
+        if image_path.exists():
+            return Image.open(image_path)
+
+
+class SegmentationJobs:
```
Comment (koosk) [inline telesto/apps/segmentation.py:69]: I am a bit worried that this part of the code will be in the public repository, but I remember this was the plan in the drawing.

File: telesto/apps/segmentation.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        logger.info("Worker thread started")
+
+        while True:
+            if job_queue:
+                job_id = job_queue.pop()
+                logger.info(f"Processing task {job_id}")
+                model_wrapper(job_id)
+                logger.info(f"Finished task {job_id}")
+
+            time.sleep(1)
```
Comment (koosk) [inline telesto/apps/segmentation.py:144]: This seems quite useless here, maybe it should be in an `else` block?

File: tests/test_apps_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    assert resp.status == falcon.HTTP_404, resp.text
+
+
+def test_segm_get(client: testing.TestClient, storage: ImageStorage):
+    array = np.array([[0, 0], [1, 1]])
+    image = Image.fromarray(array, mode="L")
+    job_id = "abc"
+    storage.save(job_id, image, is_mask=True)
+
+    resp = client.simulate_get(f"/jobs/{job_id}")
```
Comment (koosk) [inline tests/test_apps_segmentation.py:81]: It is not so clear for me how we get the result. Is the call to `predict` (possibly in `SegmentationJob`) left out on purpose for now?

**Review by koosk — COMMENTED** [review]
I left a few comments, feel free to ignore them, I don't really expect answers.
Scalability seems to be an issue, but the code is based on the plan so far that we've discussed.

**Comment by intsco** [comment]
@koosk I guess the GH UI should provide a way to review only the changes since your last review. If it's not the case, feel free to skip this review request. BTW, I addressed some and tried to answer the other your comments.

File: telesto/apps/segmentation.py (hunk 3 above)
Comment (koosk) [reply telesto/apps/segmentation.py:48]: Okay, but is it always going to run in a container or will someone try to run it on Windows?

**Review by cosmic-cortex — APPROVED** [review]

**Review by koosk — APPROVED** [review]
It was a bit hard for me to understand what is where, but I think I get it now.

File: telesto/apps/segmentation.py (hunk 3 above)
Comment (intsco) [reply telesto/apps/segmentation.py:48]: Good point, I always forget about people who code on Windows. I'll address it.

**Pull request merged**